include/seastar/core/future.hh

```diff
@@ -490,11 +491,25 @@ struct future_state_base {
         state st;
         std::exception_ptr ex;
     } _u;
+#ifdef SEASTAR_DEBUG_PROMISE
+    shard_id _owner_shard = this_shard_id();
+
+    void foreign_access_error() const noexcept;
+    void assert_shard_owner() const noexcept {
+        if (_owner_shard != this_shard_id()) {
+            foreign_access_error();
+        }
+    }
+#else
+    void assert_shard_owner() const noexcept { }
+#endif
 
```

Comment:
It's not clear that it's illegal in all cases. One might pass a ready future to a different shard (as a container of an object|exception), or similarly pass an unattached promise to a different shard. It only becomes problematic once a future/promise pair is created.

If we were starting Seastar now, we could say it's illegal, but at this time this change could break existing applications. I think we need to relax it to only apply when there's a connection between the future and the promise.

Reply:
Addressed in next version